### ISSUE TITLE
Implement LongLoRA trick for efficient tuning of long-context models

### ISSUE THREAD

**Issue by casper-hansen** [huggingface/peft#958]
### Feature request

The authors of LongLoRA explore a trick you can toggle on during training and toggle off during inference. The key takeaways are:
- LoRA perplexity deteriorates as context length increases, LongLoRA solves this and has the same perplexity as FP16 on longer contexts
- Able to train with longer contexts much faster than LoRA while keeping the same VRAM usage as LoRA

![image](https://github.com/huggingface/peft/assets/27340033/ffb6870f-28b2-4c3a-afa1-da65ea2e335a)

Paper: https://arxiv.org/pdf/2309.12307.pdf
Code: https://github.com/dvlab-research/LongLoRA

### Motivation

The key motivation is to enable a LoRA method that scales better for fine-tuning models with longer context sizes. We know this to be expensive both in compute hours and memory requirements, but LongLoRA seems to optimize for this. Using LongLoRA, we are able to train models faster and keep the perplexity of FP16 on longer context sizes.

### Your contribution

The core part of the code is written here:
https://github.com/dvlab-research/LongLoRA/blob/4cbb61b93ac8c7ea4db7ffc24a824f3fb153554d/llama_attn_replace.py#L83-L89

**Comment by BenjaminBossan** [comment]
This looks very promising, especially if it really just requires a couple of lines of extra code. Thanks for bringing this to our attention. Do you want to work on adding this?

**Comment by casper-hansen** [comment]
@BenjaminBossan I am currently occupied with a lot of other work, and I do not know where to start with implementing this in PEFT, so I would hope one of the great HF MLEs can pick this up🙏

**Comment by BenjaminBossan** [comment]
Haha, let's see if flattery does the trick for you ;-)

I was mainly asking because it was not clear to me from your post if you wanted to add it or just suggested it to be added, not to push you.

**Comment by casper-hansen** [comment]
Haha, maybe it will :)

I mainly posted here because I saw it could have a large benefit for all users looking to train LLMs with QLoRA and long contexts

**Comment by marcasty** [comment]
I'd like to add this feature if one of the great HF MLEs hasn't picked it up yet ;)

**Comment by BenjaminBossan** [comment]
We haven't started on it yet, so feel free to take a shot. Don't hesitate to ask questions or to create a draft PR early, so that we can give quick feedback.

**Comment by marcasty** [comment]
I read some code and have a few thoughts: 
1. Should there be a boolean argument to LoraConfig() in peft/tuners/lora/config.py to specify if LongLora is to be used
2. The authors of LongLoRA implemented this for Llama by updating the forward pass in transformers.models.llama.modeling_llama.LlamaAttention then passing that model to LoraConfig. It almost feels like a transformers project rather than peft

So does supporting this feature require custom LongLoRA attention implementations for each model?

**Comment by BenjaminBossan** [comment]
> 2\. The authors of LongLoRA implemented this for Llama by updating the forward pass in transformers.models.llama.modeling_llama.LlamaAttention then passing that model to LoraConfig. It almost feels like a transformers project rather than peft
> 
> So does supporting this feature require custom LongLoRA attention implementations for each model?

Interesting. In theory, we can do that in PEFT also, whether that's a good idea or not depends on what exactly is happening in the modified `forward` method.

Looking at the code [here](https://github.com/dvlab-research/LongLoRA/blob/408aef1d4b3dc93bbf83396cad7e0df7fd186fba/llama_attn_replace.py), it looks quite difficult to implement this in a generic way. In particular, these new methods (actually, functions) don't call `super().forward`, i.e. they need to replicate exactly what happens inside the method they replace. Perhaps it would be possible to extract the generic parts and put them into a solution that is not model-specific, but that would require some deep digging to figure out.

**Comment by casper-hansen** [comment]
@BenjaminBossan So in other words, it would be difficult to implement in PEFT because you have to hijack the attention module and add the appropriate calls in the right places? Another solution is for other training libraries to implement custom forward methods and implement it in those, e.g. as they do in axolotl for llama or mistral models. The problem is that this causes a lot of overhead when trying to support this feature.

**Comment by BenjaminBossan** [comment]
> So in other words, it would be difficult to implement in PEFT because you have to hijack the attention module and add the appropriate calls in the right places?

I think the answer is "it depends". We could do the same thing as in that repo, i.e. provide highly model-specific solutions and then only enable LongLoRA on those models. But I would argue that there would only be little benefit of using PEFT if it just does the exact same, narrow thing.

It would be much better to have a more generic implementation, but whether that works depends on the changes that were introduced. I cannot comment on that, because I haven't looked into the changes in detail. But if they could be generalized and abstracted, they would be great additions to PEFT.

To give a toy example, if the changes amount to adding 1 to to the output of the attention layer, that would be possible to add to PEFT in a generalized fashion. If the changes require 10 modifications in different lines in the `forward` method, that could not be generalized. Does that make sense?

> Another solution is for other training libraries to implement custom forward methods and implement it in those

Could you elaborate on how that is different? Thanks.

**Comment by casper-hansen** [comment]
Yes, I do agree. It looks like you need several modifications that are neither at the start nor the end of the `forward`. That makes it quite hard to generalize.

> Could you elaborate on how that is different? Thanks.

Training libraries like axolotl implement custom forward functions for some models in order to implement sample packing and enable features like flash attention. They also build on top of PEFT and other Huggingface libraries.

**Comment by marcasty** [comment]
the task that motivates this paper is to extend the context length for transformer language models. Does PEFT support extending the context length (I couldn't find it)? The code we're discussing is only Shift Short Attention, so just implementing Shift Short Attention would allow easy efficient long context fine-tuning but not efficient context extension. I think the Shift Short Attention implementation is the bigger fish to fry but thought this was worth mentioning to appreciate the scope of this feature

**Comment by BenjaminBossan** [comment]
> Training libraries like axolotl implement custom forward functions for some models in order to implement sample packing and enable features like flash attention. They also build on top of PEFT and other Huggingface libraries.

Okay, so your suggestion is to offload the task of implementing the custom forward methods to the training libs? Yes, there is a bit of a blurry line if these types of changes are in the scope of PEFT or not. I'd say yes, as it's not strictly related to specific training methods, but it could be done either way.

> Does PEFT support extending the context length (I couldn't find it)?

No, the idea would be for PEFT to make the modification to the layers, the training itself on longer contexts would be up to the user.

**Comment by teknium1** [comment]
dead?

**Comment by matteoguarrera** [comment]
> I'd like to add this feature if one of the great HF MLEs hasn't picked it up yet ;)

You still working on that? I wanna help

**Comment by seanxuu** [comment]
> > 如果一个伟大的 HF MLE 还没有拿起它，我想添加这个功能;)
> 
> 你还在努力吗？我想帮忙

same question

**Comment by Gauravbhai4** [comment]
can we implement this on longformer model
